x-pack/plugin/esql/arrow/src/main/java/org/elasticsearch/xpack/esql/arrow/BlockConverter.java

```diff
@@ -0,0 +1,477 @@
+            bufs.add(BlockConverter.dummyArrowBuf(dataVectorLength(block)));
+
+            bufWriters.add(out -> {
+                if (block.areAllValuesNull()) {
+                    return 0;
+                }
+
+                // TODO could we "just" get the memory of the array and dump it?
+                BytesRef scratch = new BytesRef();
+                long length = 0;
+                for (int i = 0; i < block.getPositionCount(); i++) {
+                    BytesRef v = block.getBytesRef(i, scratch);
+
+                    out.write(v.bytes, v.offset, v.length);
+                    length += v.length;
+                }
+                return length;
+            });
+        }
+
+        private static int offsetVectorLength(BytesRefBlock block) {
+            return Integer.BYTES * (block.getPositionCount() + 1);
+        }
+
+        private int dataVectorLength(BytesRefBlock block) {
+            if (block.areAllValuesNull()) {
+                return 0;
+            }
+
+            // TODO we can probably get the length from the vector without all this sum
+
+            int length = 0;
+            BytesRef scratch = new BytesRef();
+            for (int i = 0; i < block.getPositionCount(); i++) {
+                BytesRef v = block.getBytesRef(i, scratch);
+                length += v.length;
+            }
+            return length;
+        }
+    }
+
+    /**
+     * Conversion of ByteRefs where each value is itself converted to a different format.
+     */
+    public static class TransformedBytesRef extends BytesRefConverter {
+
+        private final BiFunction<BytesRef, BytesRef, BytesRef> valueConverter;
+
+        /**
+         *
+         * @param esqlType ESQL type name
+         * @param minorType Arrow type
+         * @param valueConverter a function that takes (value, scratch) input parameters and returns the transformed value
+         */
+        public TransformedBytesRef(String esqlType, Types.MinorType minorType, BiFunction<BytesRef, BytesRef, BytesRef> valueConverter) {
+            super(esqlType, minorType);
+            this.valueConverter = valueConverter;
+        }
+
+        @Override
+        public void convert(Block b, List<ArrowBuf> bufs, List<BufWriter> bufWriters) {
+            BytesRefBlock block = castBlock(b, BytesRefBlock.class);
+            BytesRefBlock transformed = transformValues(block);
+            super.convert(transformed, bufs, bufWriters);
+        }
+
+        /**
+         * Creates a new BytesRefBlock by applying the value converter to each non null and non empty value
+         */
+        private BytesRefBlock transformValues(BytesRefBlock block) {
+            BlockFactory blockFactory = BlockFactory.getInstance(
+                // FIXME: are these parameters ok here?
+                // FIXME: should the resulting block be closed after use?
+                new NoopCircuitBreaker("esql-arrow"),
+                BigArrays.NON_RECYCLING_INSTANCE
```

Comment:
This is fine for a prototype but it's a blocker to stop switch. There are three choices I think that are ok:
1. Push the converter into the superclass and do it optionally, right as you dump into the wire. This is probably the best way because it involves the least copying and memory management. It might be the slowest because the converter will go megamorphic pretty quick. But it's shoveling a fair number of bits around so it's probably fine. Probably.
2. Convert the whole block up front in a tight loop with memory management.
3. Option 1 but with the generics reified out by hand - java doesn't do it for you so you need to do it with your own fingers. Or, in our case, we'd do it with stringtemplate. That's almost certainly the choice that's fastest at runtime. But I don't know that it matters for this.

Either way, this is fine for now.

Reply:
I don't know much about memory management in ES. Is there some explanation somewhere or some reference code I could get inspiration from?

Reply:
It's more memory management int he compute engine than the rest of ES. The short version is that you can use `BlockFactory` to make a builder like:
```
try (IntBlockBuilder builder = block.blockFactory().newIntBlockBuilder(block.getPositionCount())) {
  for... { builder.appendInt(i); }
  try (IntBlock block = builder.build()) {
    super.render(block);
  }
}
```
That'll account for the new block using the same circuit breakers as the original block. It *could* be fairly large.

Something like `RoundIntEvaluator` does something similar.
}

Reply:
Added 54fa11f52dba3332abffd2cf06e5d95a47283af2 to reuse the source block's factory and inherits its circuit breaker.